File: go/VigenereCipher.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
)

type VigenereCipher struct {
	key string
}

func (vc *VigenereCipher) normalizeKey(length int) {
	keyLength := len(vc.key)
	repeat := length / keyLength
	left := length % keyLength
	normalizedKey := ""

	for i := 0; i < repeat; i++ {
		normalizedKey += vc.key
	}

	normalizedKey += vc.key[:left]

	vc.key = normalizedKey
}

func (vc *VigenereCipher) encode(plainText string) string {
	if len(vc.key) < len(plainText) {
		vc.normalizeKey(len(plainText))
	}

	var chr rune = 0

	result := []rune{}

	for i, char := range plainText {
		if char >= 97 && char <= 122 {
			chr = ((char - 97) + (rune(vc.key[i]) - 97)) % 26 + 97

			result = append(result, chr)
		} else {
			result = append(result, char)
		}
	}

	return string(result)
}

func (vc *VigenereCipher) decode(cipherText string) string {
	if len(vc.key) < len(cipherText) {
		vc.normalizeKey(len(cipherText))
	}

	var chr rune = 0

	result := []rune{}

	for i, char := range cipherText {
		if char >= 97 && char <= 122 {
			chr = ((char - 97) - (rune(vc.key[i]) - 97)) % 26

			if chr < 0 {
				chr += 26
			}

			chr += 97

			result = append(result, chr)
		} else {
			result = append(result, char)
		}
	}

	return string(result)
}

func main() {
	fmt.Println(`VIGENERE CIPHER
1. Encode
2. Decode
	`)

	fmt.Println("Choose an option:")

	option := 0

	fmt.Scanln(&option)

	fmt.Println("Enter text:")

	text, _ := bufio.NewReader(os.Stdin).ReadString('\n')

	text = strings.ToLower(text)

	fmt.Println("Enter key:")

	key, _ := bufio.NewReader(os.Stdin).ReadString('\n')

	key = strings.TrimSpace(key)

	vc := VigenereCipher{ key }

	if option == 1 {
		result := vc.encode(text)

		fmt.Printf("Cipher text: %v", result)
	}

	if option == 2 {
		result := vc.decode(text)

		fmt.Printf("Plain text: %v", result)
	}
}

File: php/VigenereCipher.php
<?php
declare(strict_types=1);

class VigenereCipher
{
  public string $key;

  public function __construct(string $key)
  {
    $this->key = $key;
  }

  function normalizeKey(int $length)
  {
    $keyLength = strlen($this->key);
    $repeat = intval(floor($length / $keyLength));
    $left = $length % $keyLength;
    $normalizedKey = str_repeat($this->key, $repeat);
    $normalizedKey .= substr($this->key, 0, $left);
    $this->key = $normalizedKey;
  }

  public function encode(string $plainText)
  {
    $length = strlen($plainText);

    if (strlen($this->key) < $length) {
      $this->normalizeKey($length);
    }

    $result = "";
    $char = 0;

    for ($i = 0; $i < $length; $i++) {
      $code = ord($plainText[$i]);

      if ($code >= 97 && $code <= 122) {
        $char = (($code - 97) + (ord($this->key[$i]) - 97)) % 26 + 97;

        $result .= chr($char);
      } else {
        $result .= $plainText[$i];
      }
    }

    return $result;
  }

  public function decode(string $cipherText)
  {
    $length = strlen($cipherText);

    if (strlen($this->key) < $length) {
      $this->normalizeKey($length);
    }

    $result = "";
    $char = 0;

    for ($i = 0; $i < $length; $i++) {
      $code = ord($cipherText[$i]);

      if ($code >= 97 && $code <= 122) {
        $char = (($code - 97) - (ord($this->key[$i]) - 97)) % 26;

        if ($char < 0) $char += 26;

        $char += 97;

        $result .= chr($char);
      } else {
        $result .= $cipherText[$i];
      }
    }

    return $result;
  }
}

echo <<<EOD
VIGENERE CIPHER
1. Encode
2. Decode
EOD;

echo "\n\nChoose an option:\n";

$option = intval(trim(fgets(fopen("php://stdin", "r"))));

echo "Enter text:\n";

$text = trim(fgets(fopen("php://stdin", "r")));

$text = strtolower($text);

echo "Enter key:\n";

$key = trim(fgets(fopen("php://stdin", "r")));

$vc = new VigenereCipher($key);

if ($option === 1) {
  $result = $vc->encode($text);

  echo "Cipher text: $result";
}

if ($option === 2) {
  $result = $vc->decode($text);

  echo "Plain text: $result";
}